mss/L2Cache.hpp

```diff
@@ -0,0 +1,376 @@
+
+        sparta::DataOutPort<olympia::InstPtr> out_biu_req_
+            {&unit_port_set_, "out_l2cache_biu_req"};
+
+        sparta::DataOutPort<olympia::InstPtr> out_l2cache_icache_resp_
+            {&unit_port_set_, "out_l2cache_icache_resp"};
+
+        sparta::DataOutPort<olympia::InstPtr> out_l2cache_dcache_resp_
+            {&unit_port_set_, "out_l2cache_dcache_resp"};
+
+        sparta::DataOutPort<uint32_t> out_l2cache_icache_ack_
+            {&unit_port_set_, "out_l2cache_icache_ack"};
+
+        sparta::DataOutPort<uint32_t> out_l2cache_dcache_ack_
+            {&unit_port_set_, "out_l2cache_dcache_ack"};
+
+
+        ////////////////////////////////////////////////////////////////////////////////
+        // Internal States
+        ////////////////////////////////////////////////////////////////////////////////
+
+        
+        using CacheRequestQueue = std::vector<olympia::InstPtr>;
+
+        // Buffers for the incoming requests from DCache and ICache
+        CacheRequestQueue dcache_req_queue_;
+        CacheRequestQueue icache_req_queue_;
+
+        const uint32_t dcache_req_queue_size_;
+        const uint32_t icache_req_queue_size_;
+
+        // Buffers for the outgoing requests from L2Cache
+        CacheRequestQueue biu_req_queue_;
+
+        const uint32_t biu_req_queue_size_;
+
+        // Buffers for the incoming resps from BIU
+        CacheRequestQueue biu_resp_queue_;
+
+        const uint32_t biu_resp_queue_size_;
+
+        // Buffers for the outgoing resps to DCache and ICache
+        CacheRequestQueue dcache_resp_queue_;
+        CacheRequestQueue icache_resp_queue_;
+
+        const uint32_t dcache_resp_queue_size_;
+        const uint32_t icache_resp_queue_size_;
+
+        // Channels enum
+        enum class Channel : uint32_t
+        {
+            NO_ACCESS = 0,
+            __FIRST = NO_ACCESS,
+            BIU,
+            ICACHE,
+            DCACHE,
+            NUM_CHANNELS,
+            __LAST = NUM_CHANNELS
+        };
+        
+
+        // Cache Pipeline
+        class PipelineStages {
+            public:
+                PipelineStages(uint32_t latency) : NUM_STAGES(latency),
+                                                   HIT_MISS_HANDLING(NUM_STAGES - 1),
+                                                   CACHE_LOOKUP(HIT_MISS_HANDLING - 1) {}
+
+            const uint32_t NUM_STAGES;
+            const uint32_t HIT_MISS_HANDLING;
+            const uint32_t CACHE_LOOKUP;
+            const uint32_t NO_ACCESS = 0;
+        };
+        
+        using L2ArchUnit = olympia::MemoryAccessInfo::ArchUnit;
+        using L2CacheState = olympia::MemoryAccessInfo::CacheState;
+        using L2CachePipeline = sparta::Pipeline<olympia::MemoryAccessInfoPtr>;
+        
+        PipelineStages stages_;
+        L2CachePipeline l2cache_pipeline_;
+        
+        sparta::Queue<olympia::MemoryAccessInfoPtr> pipeline_req_queue_;
+        const uint32_t pipeline_req_queue_size_;
```

Comment:
Is this used?  If not, feel free to delete it.  You can query the size from the `sparta::Queue`